examples/nlp/gpt/train_gpt_spin.py

```diff
@@ -0,0 +1,186 @@
+    exp_manager(trainer, cfg.exp_manager)
+    logger = CustomLoggerWrapper(trainer.loggers)
+
+    ptl_model = load_from_nemo(
+        MegatronGPTSPINModel,
+        cfg.model,
+        trainer,
+        strict=True,
+        restore_path=cfg.pretrained_checkpoint.restore_from_path,
+    )
+
+    with open_dict(cfg):
+        # overwrite the model config with the config from the checkpoint
+        cfg.model.encoder_seq_length = ptl_model.cfg.encoder_seq_length
+
+    if ptl_model.ref_policy_state_dict is None:
+        ref_policy_state_dict = retrieve_model_state_dict_in_cpu(
+            ptl_model, megatron_amp_O2=cfg.model.get("megatron_amp_O2", False)
+        )
+        ptl_model.ref_policy_state_dict = ref_policy_state_dict
+
+    # pull values from checkpoint
+    trainer_restore_path = trainer.ckpt_path
+
+    # TODO: log this restore path
+    if trainer_restore_path is not None:
+        custom_trainer_state_dict = retrieve_custom_trainer_state_dict(trainer)
+        consumed_samples = custom_trainer_state_dict["consumed_samples"]
+    else:
+        custom_trainer_state_dict = None
+        consumed_samples = 0
+
+    init_distributed(trainer, ptl_model, cfg.model.get("transformer_engine", False))
+
+    train_data_cfg = cfg.model.data.train_ds
+    val_data_cfg = cfg.model.data.validation_ds
+
+    if cfg.model.data.get("sample", False):
+        # if it is negative, num_samples is None
+        if cfg.trainer.spin.max_steps < 0:
+            num_samples = None
+        else:
+            num_samples = cfg.trainer.spin.max_steps * train_data_cfg.global_batch_size
+    else:
+        num_samples = None
+    train_ds = build_sft_dataset(
+        train_data_cfg,
+        ptl_model.tokenizer,
+        num_samples,
+        answer_only_loss=True,
+        is_chat=cfg.model.data.chat,
+        special_tokens=cfg.model.data.chat_prompt_tokens,
+    )
+
+    if cfg.model.data.get("sample", False):
+        num_samples = cfg.trainer.spin.limit_val_batches * val_data_cfg.global_batch_size
+    else:
+        num_samples = None
+    validation_ds = build_sft_dataset(
+        val_data_cfg,
+        ptl_model.tokenizer,
+        num_samples,
+        answer_only_loss=True,
+        is_chat=cfg.model.data.chat,
+        special_tokens=cfg.model.data.chat_prompt_tokens,
+    )
+
+    eos_id = ptl_model.tokenizer.eos_id
+
+    # collate fn to pad to the max seq length in the batch
+    collate_fn = partial(
+        spin_custom_collate,
+        eos_id=eos_id,
+        reset_position_ids=cfg.model.data.get("reset_position_ids", False),
+        reset_attention_mask=cfg.model.data.get("reset_attention_mask", False),
+        eod_mask_loss=cfg.model.data.get("eod_mask_loss", False),
+    )
+
+    train_dataloader = build_dataloader(
+        cfg=cfg,
+        dataset=train_ds,
+        consumed_samples=consumed_samples,
+        mbs=cfg.model.micro_batch_size,
+        gbs=cfg.model.global_batch_size,
+        collate_fn=collate_fn,
```

Comment:
Correct, yes, good catch, I have changed everything now to be in line with how SFT does it

Reply:
As far as I can tell you changed the handling of micro / global batch sizes, but `collate_fn` for the training set is still your `spin_custom_collate` instead of `train_ds.collate_fn`. Is there a specific reason we need your custom function?

Reply:
Yes, for SPIN, we need custom output from each batch coming from the dataloader which `train_ds.collate_fn` does not give us. These custom outputs are necessary for SPIN to work correctly, hence we need `spin_custom_collate`. You can look at both functions to see the differences between them

Reply:
Got it, thanks. I had noticed they were quite different and originally it wasn't obvious to me why, especially since you were using `validation_ds.collate_fn`. But I get it now -- validation is very different from training in your SPIN implementation since it's using an SFT loss, which explains this discrepancy.